core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -1133,23 +1133,27 @@ private[kafka] class Processor(
                 expiredConnectionsKilledCount.record(null, 1, 0)
               } else {
                 val connectionId = receive.source
-                val context = new RequestContext(header, connectionId, channel.socketAddress, Optional.of(channel.socketPort()),
-                  channel.principal, listenerName, securityProtocol, channel.channelMetadataRegistry.clientInformation,
-                  isPrivilegedListener, channel.principalSerde)
-
-                val req = new RequestChannel.Request(processor = id, context = context,
-                  startTimeNanos = nowNanos, memoryPool, receive.payload, requestChannel.metrics, None)
-
+                var requestAndSize: Option[RequestAndSize] = None
                 // KIP-511: ApiVersionsRequest is intercepted here to catch the client software name
                 // and version. It is done here to avoid wiring things up to the api layer.
                 if (header.apiKey == ApiKeys.API_VERSIONS) {
-                  val apiVersionsRequest = req.body[ApiVersionsRequest]
+                  val result = RequestContext.staticParseRequest(header, receive.payload, connectionId, listenerName, channel.principal)
+                  val apiVersionsRequest = result.request.asInstanceOf[ApiVersionsRequest]
                   if (apiVersionsRequest.isValid) {
                     channel.channelMetadataRegistry.registerClientInformation(new ClientInformation(
                       apiVersionsRequest.data.clientSoftwareName,
                       apiVersionsRequest.data.clientSoftwareVersion))
```

Comment:
Could we update `context#clientInformation` directly?

Reply:
What I mean is that `context#clientInformation` gets updated when the request is related to `API_VERSIONS`.

Reply:
Hi @chia7712, thanks for the suggestion. I updated code. Could you take a look again? Thanks.